File: 2019/src/day01.dart
import '../../shared/dart/src/utils.dart';

/// Puzzle description: https://adventofcode.com/2019/day/1

const bool DEBUG = false;

int recursiveFuelRequired(int mass) {
  var fuel = fuelRequired(mass);
  return (fuel <= 0) ? 0 : fuel + recursiveFuelRequired(fuel);
}

//alternative to above using a while loop
int loopFuelRequired(int mass) {
  var result = 0;
  var calcMass = mass;

  while (true) {
    var fuel = fuelRequired(calcMass);
    if (fuel <= 0) break;
    result += fuel;
    calcMass = fuel;
  }

  if (DEBUG) print('Total for $mass: $result');
  return result;
}

int fuelRequired(int mass) {
  var result = (mass / 3).floor() - 2;
  if (DEBUG) print('$mass: $result');
  return result;
}

int part1(String header, List<int> input) {
  printHeader(header);
  // return input.fold(0, (a, b) => a + fuelRequired(b));
  // return input.map((f) => fuelRequired(f)).sum; //alternative to above
  return input.fold2(fuelRequired) as int; //another alt
}

int part2(String header, List<int> input) {
  printHeader(header);

  return input.fold2(recursiveFuelRequired) as int; //another alt
  // return input.fold(0, (a, b) => a + recursiveFuelRequired(b));
  // return input.fold(0, (a, b) => a + loopFuelRequired(b));
}

void main(List<String> arguments) {
  var testInput = FileUtils.asInt('../data/day01-test.txt');
  var mainInput = FileUtils.asInt('../data/day01.txt');

  assertEqual(part1('01 test part 1', testInput), 34241);
  assertEqual(part2('01 test part 2', testInput), 51316);

  //Answer:3246455
  printAndAssert(part1('01 part 1', mainInput),3246455);
  //Answer:4866824
  printAndAssert(part2('01 part 2', mainInput),4866824);
}

File: 2015/src/day01.py
from utils import *
import os
import sys
sys.path.append(os.path.relpath("../../shared/python"))

# Puzzle description: https://adventofcode.com/2015/day/1


def read_input(input_file):
    input = read_file_str(input_file, True)
    return input[0]


def part1(input):
    return input.count('(') - input.count(')')


def part2(input):
    floor = 0
    for i in range(len(input)):
        if input[i] == '(':
            floor += 1
        elif input[i] == ')':
            floor -= 1

        if floor == -1:
            return i+1

    return 0


def main():
    input = read_input("../data/day01.txt")

    assert (res := part1('(())')) == 0, f'Actual: {res}'
    assert (res := part1('))(((((')) == 3, f'Actual: {res}'

    print(f'Part 1 {part1(input)}')  # 138

    assert (res := part2(')')) == 1, f'Actual: {res}'
    assert (res := part2('()())')) == 5, f'Actual: {res}'

    print(f'Part 2 {part2(input)}')  # 1771


if __name__ == '__main__':
    main()
